python/ray/serve/backend_state.py

```diff
@@ -36,119 +35,208 @@ class ReplicaState(Enum):
                 name=self._actor_name,
                 lifetime="detached" if self._detached else None,
                 max_restarts=-1,
                 max_task_retries=-1,
-                placement_group=self._placement_group,
+                placement_group=pg,
                 **backend_info.replica_config.ray_actor_options).remote(
                     self._backend_tag, self._replica_tag,
                     backend_info.replica_config.init_args,
                     backend_info.backend_config, self._controller_name)
+        self._startup_obj_ref = actor_handle.ready.remote()
+
+    def check_ready(self) -> bool:
+        ready, _ = ray.wait([self._startup_obj_ref], timeout=0)
+        return len(ready) == 1
+
+    def resource_requirements(
+            self) -> Tuple[Dict[str, float], Dict[str, float]]:
+        """Returns required and currently available resources.
+
+        Only resources with nonzero requirements will be included in the
+        required dict and only resources in the required dict will be
+        included in the available dict (filtered for relevance).
+        """
+        required = {k: v for k, v in self._actor_resources.items() if v > 0}
+        available = {
+            k: v
+            for k, v in ray.available_resources().items() if k in required
+        }
+        return required, available
+
+    def graceful_stop(self) -> None:
+        # NOTE: the replicas may already be stopped if we failed
+        # after stopping them but before writing a checkpoint.
+        if self._stopped:
+            return
+
+        try:
+            handle = ray.get_actor(self._actor_name)
+            self._drain_obj_ref = handle.drain_pending_queries.remote()
+        except ValueError:
+            self._stopped = True
+
+    def check_stopped(self) -> bool:
+        if self._stopped:
+            return True
+
+        try:
+            ray.get_actor(self._actor_name)
+            ready, _ = ray.wait([self._drain_obj_ref], timeout=0)
+            self._stopped = len(ready) == 1
+        except ValueError:
+            self._stopped = True
+
+        return self._stopped
+
+    def force_stop(self):
+        try:
+            ray.kill(ray.get_actor(self._actor_name))
+        except ValueError:
+            pass
+        try:
+            ray.util.remove_placement_group(
+                ray.util.get_placement_group(self._placement_group_name))
+        except ValueError:
+            pass
+
+
+class BackendReplica:
+    """Manages state transitions for backend replicas.
+
+    This is basically a checkpointable lightweight state machine.
+    """
+
+    def __init__(self, controller_name: str, detached: bool,
+                 replica_tag: ReplicaTag, backend_tag: BackendTag):
+        self._actor = ActorReplicaWrapper(
+            format_actor_name(replica_tag, controller_name), detached,
+            controller_name, replica_tag, backend_tag)
```

Comment:
This is the main change -- I shoved the actor-related API calls into this class so we can mock it out and not actually spin up actors